Feature: OTEL service daemon for ECS on EC2 container instances.

frank-ang opened the pull request:
# Description

Terraform module provisioning OTEL agent collector as an ECS service daemon. Supports EC2 container instances.

Feature request: [CDS-762 ](https://coralogix.atlassian.net/browse/CDS-762)Create OTEL ecs-ec2 Terraform.

# How Has This Been Tested?

* Environment Setup:
  * Create New AWS VPC, ECS Cluster on EC2 container instances. 
  * Deploy test nginx service.
  * Open a terminal with valid AWS session credentials and Coralogix key as ``` TF_VAR_private_key```
  * Review ```terraform-coralogix-aws/tests/ecs-ec2/ecs-ec2.tf```, edit ```region``` and ```coralogix_region``` accordingly to your environment.
  * Deploy ```ecs-ec2``` Terraform module:
    ```terraform
    cd terraform-coralogix-aws/tests/ecs-ec2
    terraform init
    terraform plan
    terraform apply -auto-approve
    ```
* Testing: 
  * Verify ECS service named ```coralogix-otel-agent``` runs as a daemon task on all EC2 instances.
  * Send requests to nginx container endpoint.
  * Verify Coralogix console shows nginx logs and container metrics

# Checklist:
- [ ✅ ] I have updated the relevant example in the examples directory for the module.
- [ ✅ ] I have updated the relevant test file for the module.
- [ ✅ ] This change does not affect module (e.g. it's readme file change)

royfur commented on examples/ecs-ec2/variables.tf:
```diff
@@ -0,0 +1,44 @@
+  default     = "default"
+}
+
+variable "private_key" {
```
Should be changed to api key

frank-ang replied:
Its better to retain using "private_key", for consistency with: 
* Coralogix exporter config variable name: "private_key"
* Coralogix otel collector docker env var: "PRIVATE_KEY"
* AWS Cloudformation ECS-EC2 param name: "PrivateKey"
* AWS Cloudformation ECS-Fargate param name: "PrivateKey"

royfur commented on examples/ecs-ec2/variables.tf:
```diff
@@ -0,0 +1,44 @@
+  default     = 256
+}
+
+variable "coralogix_region" {
```
Need to support also custom region

frank-ang replied:
Will test across multiple regions.

royfur replied:
The custom region is basically address that is not in the mapping list
It is for use cases of private cluster or new cluster ....
So we need to be able to put the coralogix endpoint

frank-ang replied:
```custom_domain``` parameter added.

royfur commented on examples/ecs-ec2/variables.tf:
```diff
@@ -0,0 +1,44 @@
+}
+
+variable "image_version" {
+  description = "The Coralogix Open Telemetry Distribution Image Version/Tag. Defaults to \"latest\". See: https://hub.docker.com/r/coralogixrepo/coralogix-otel-collector/tags"
```
We don't use latest should be selected manually and we don't need to have default

royfur commented on examples/ecs-ec2/variables.tf:
```diff
@@ -0,0 +1,44 @@
+  sensitive   = true
+}
+
+variable "metrics" {
```
The description is not good, if try means that we will collect beside logs and traces also metrics

frank-ang replied:
This behavior is consistent with the Cloudformation version, that uses ```metrics``` as a boolean toggle for metrics on/off. Logs and Traces are always on and we don't give the user the choice. 
Parameter description has been improved.

frank-ang commented:
Additional capabilities pending investigation:
* OTEL container health check.
* Test for user-defined OTEL Config file.
* Integration test across multiple regions and Coralogix endpoints.
* Support for Windows

daidokoro reviewed (approved):
LGTM, only need to resolve the recommendations that @royfur made.

frank-ang commented:
Implemented changes requested by @royfur , including:
* Improved variables:
    * ```custom_domain``` parameter to override ```coralogix_region``` parameter
    * ```api_key``` instead of ```private_key```
* Improved test configuration. Added test readme. 
* Tested multi-region (SG and IN), custom_domain.

frank-ang commented:
> Additional capabilities pending investigation:
> 
> * OTEL container health check.
> * Test for user-defined OTEL Config file.
> * Integration test across multiple regions and Coralogix endpoints.
> * Support for Windows

Windows will be implemented in a separate Terraform template with a separate PR.

frank-ang commented:
Added my GPG key and retro-signed commits 1307029..2e9167 for this PR.
